File: clients/beebium-typescript-client/src/extension_ui.ts
/**
 * Wrapper for ExtensionUiService.
 *
 * The Extension UI framework lets any extension declare a typed control
 * tree (the View) which the server streams to clients. Clients render
 * natively and post user actions back through Dispatch. This module
 * wraps the gRPC SubscribeView / Dispatch RPCs and converts the proto
 * types into ergonomic interfaces.
 *
 * See docs/discussion/extension-ui-architecture.md for the design.
 *
 * Usage:
 *     // One-shot: consume the first view of the piconet panel.
 *     for await (const view of bbc.extensionUi.subscribeView("piconet")) {
 *         for (const child of view.root.group?.children ?? []) {
 *             console.log(child.id, child.kind);
 *         }
 *         break;
 *     }
 *
 *     // Toggle the piconet "Enabled" control.
 *     await bbc.extensionUi.dispatch("piconet", "mode_toggle", view.revision, false);
 *
 *     // Background: receive view updates as they arrive.
 *     const handle = bbc.extensionUi.startBackgroundSubscription(
 *         "piconet",
 *         (v) => console.log("revision", v.revision),
 *     );
 *     // ...
 *     handle.stop();
 */

import type {
    ExtensionUiServiceClient,
    View as ProtoView,
    Control as ProtoControl,
    DispatchRequest as ProtoDispatchRequest,
    DispatchResponse as ProtoDispatchResponse,
} from "./generated/extension_ui.js";
import { promisify } from "./call-utils.js";
import { BackgroundStreamHandle, toAsyncIterable } from "./stream-utils.js";

/**
 * Semantic state for an Indicator control.
 *
 * Renderers map this to whatever colour / icon convention their
 * platform prefers; the server reports semantics, not presentation.
 */
export enum IndicatorState {
    UNKNOWN = 0,
    OK = 1,
    WARN = 2,
    ERROR = 3,
}

/** Tag for the seven control primitives plus the structural Group. */
export enum ControlKind {
    LABEL = "label",
    INDICATOR = "indicator",
    TOGGLE = "toggle",
    BUTTON = "button",
    CHOICE = "choice",
    TEXT_INPUT = "text_input",
    GROUP = "group",
    UNSET = "unset",
}

export interface Label {
    text: string;
    /**
     * Optional muted caption shown alongside the primary text. Used
     * for provenance / status / size metadata that's logically
     * attached to the primary line. Empty string means no caption.
     */
    secondaryText: string;
}

export interface Indicator {
    state: IndicatorState;
    text: string;
}

export interface Toggle {
    label: string;
    value: boolean;
}

export interface Button {
    label: string;
    enabled: boolean;
}

export interface Choice {
    label: string;
    options: readonly string[];
    selectedIndex: number;
}

export interface TextInput {
    label: string;
    value: string;
    placeholder: string;
}

/** A nested group of controls. label is undefined for unlabelled groups. */
export interface Group {
    label: string | undefined;
    children: readonly Control[];
}

/**
 * A control in the View tree.
 *
 * Exactly one of label, indicator, toggle, button, choice, textInput,
 * group is populated, indicated by the kind discriminator.
 */
export interface Control {
    id: string;
    kind: ControlKind;
    label?: Label;
    indicator?: Indicator;
    toggle?: Toggle;
    button?: Button;
    choice?: Choice;
    textInput?: TextInput;
    group?: Group;
}

/** A complete view of an extension's UI at one revision. */
export interface View {
    extensionId: string;
    revision: number;
    root: Control;
}

/** Return value from Extension UI dispatch. */
export interface DispatchResult {
    accepted: boolean;
    error: string;
}

function convertControl(proto: ProtoControl | undefined): Control {
    if (!proto) {
        return { id: "", kind: ControlKind.UNSET };
    }
    if (proto.label !== undefined) {
        return {
            id: proto.id,
            kind: ControlKind.LABEL,
            label: {
                text: proto.label.text,
                secondaryText: proto.label.secondaryText,
            },
        };
    }
    if (proto.indicator !== undefined) {
        return {
            id: proto.id,
            kind: ControlKind.INDICATOR,
            indicator: {
                state: proto.indicator.state as number as IndicatorState,
                text: proto.indicator.text,
            },
        };
    }
    if (proto.toggle !== undefined) {
        return {
            id: proto.id,
            kind: ControlKind.TOGGLE,
            toggle: {
                label: proto.toggle.label,
                value: proto.toggle.value,
            },
        };
    }
    if (proto.button !== undefined) {
        return {
            id: proto.id,
            kind: ControlKind.BUTTON,
            button: {
                label: proto.button.label,
                enabled: proto.button.enabled,
            },
        };
    }
    if (proto.choice !== undefined) {
        return {
            id: proto.id,
            kind: ControlKind.CHOICE,
            choice: {
                label: proto.choice.label,
                options: [...proto.choice.options],
                selectedIndex: proto.choice.selectedIndex,
            },
        };
    }
    if (proto.textInput !== undefined) {
        return {
            id: proto.id,
            kind: ControlKind.TEXT_INPUT,
            textInput: {
                label: proto.textInput.label,
                value: proto.textInput.value,
                placeholder: proto.textInput.placeholder,
            },
        };
    }
    if (proto.group !== undefined) {
        const children = proto.group.controls.map(convertControl);
        return {
            id: proto.id,
            kind: ControlKind.GROUP,
            group: {
                label: proto.group.label,
                children,
            },
        };
    }
    return { id: proto.id, kind: ControlKind.UNSET };
}

function convertView(proto: ProtoView): View {
    return {
        extensionId: proto.extensionId,
        revision: proto.viewRevision,
        root: convertControl(proto.root),
    };
}

/** Handle for a background view subscription. */
export class SubscriptionHandle {
    private readonly handle: BackgroundStreamHandle;

    constructor(handle: BackgroundStreamHandle) {
        this.handle = handle;
    }

    /** Signal the background subscription to stop. */
    stop(): void {
        this.handle.cancel();
    }

    get isRunning(): boolean {
        return this.handle.isRunning;
    }
}

/**
 * Wraps ExtensionUiService.
 *
 * Use subscribeView to iterate over View updates for a named
 * extension, and dispatch to post a user action back. The server
 * delivers the result of an action as the next yielded View on the
 * subscription, not in the dispatch return -- DispatchResult.accepted
 * only reports whether the framework's validation gauntlet (extension
 * exists, control id known, payload type matches, revision current)
 * passed.
 */
export class ExtensionUi {
    private readonly stub: ExtensionUiServiceClient;

    constructor(stub: ExtensionUiServiceClient) {
        this.stub = stub;
    }

    /**
     * Open a server-stream of View updates for one extension.
     *
     * The first View arrives immediately; subsequent Views arrive as
     * the extension's state changes. The stream ends with NOT_FOUND
     * if the named extension does not exist or has no UI.
     */
    async *subscribeView(extensionId: string): AsyncGenerator<View, void, undefined> {
        const stream = this.stub.subscribeView({ extensionId });
        for await (const proto of toAsyncIterable(stream)) {
            yield convertView(proto);
        }
    }

    /**
     * Subscribe in the background; invoke callback on each View.
     *
     * Returns a handle whose stop() cancels the stream. Stream errors
     * (e.g. server disconnect, NOT_FOUND) terminate delivery silently
     * -- check handle.isRunning if the caller needs to detect that.
     */
    startBackgroundSubscription(
        extensionId: string,
        callback: (view: View) => void,
    ): SubscriptionHandle {
        const stream = this.stub.subscribeView({ extensionId });
        const handle = new BackgroundStreamHandle(
            stream,
            (message: unknown) => callback(convertView(message as ProtoView)),
        );
        return new SubscriptionHandle(handle);
    }

    /**
     * Post a user action back to the server.
     *
     * The payload is type-dispatched to match the addressed control's
     * expected variant:
     *
     *   * boolean                  -> Toggle
     *   * string                   -> TextInput
     *   * non-negative integer     -> Choice (selected index)
     *   * null / undefined         -> Button (no payload)
     *
     * Returns { accepted: false, error } if the framework's validation
     * gauntlet rejects the request (stale revision, unknown control,
     * payload-type mismatch, unknown extension). The actual mutation of
     * the extension's state is observable on the SubscribeView stream
     * as the next pushed View, not in the return value.
     *
     * @throws RangeError if the payload is a negative number.
     * @throws TypeError if the payload is a non-integer number or any
     *     other unsupported type.
     */
    async dispatch(
        extensionId: string,
        controlId: string,
        viewRevision: number,
        payload: boolean | string | number | null | undefined = undefined,
    ): Promise<DispatchResult> {
        const request: ProtoDispatchRequest = {
            extensionId,
            controlId,
            viewRevision,
        };

        if (payload === null || payload === undefined) {
            // leave the oneof unset (Button)
        } else if (typeof payload === "boolean") {
            request.boolValue = payload;
        } else if (typeof payload === "string") {
            request.stringValue = payload;
        } else if (typeof payload === "number") {
            if (!Number.isInteger(payload)) {
                throw new TypeError(
                    `dispatch payload must be boolean, string, integer, null, or undefined; got non-integer number ${payload}`,
                );
            }
            if (payload < 0) {
                throw new RangeError(
                    `Choice index payload must be non-negative; got ${payload}`,
                );
            }
            request.indexValue = payload;
        } else {
            throw new TypeError(
                `dispatch payload must be boolean, string, integer, null, or undefined; got ${typeof payload}`,
            );
        }

        const response = await promisify<ProtoDispatchRequest, ProtoDispatchResponse>(
            this.stub as unknown as Record<string, Function>,
            "dispatch",
            request,
        );
        return { accepted: response.accepted, error: response.error };
    }
}

File: clients/beebium-python-client/src/beebium/client/extension_ui.py
# Copyright 2026 Robert Smallshire <robert@smallshire.org.uk>
#
# This file is part of Beebium.
#
# Beebium is free software: you can redistribute it and/or modify it under the terms of the
# GNU General Public License as published by the Free Software Foundation, either version 3 of the
# License, or (at your option) any later version. Beebium is distributed in the hope that it will
# be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
# FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
# You should have received a copy of the GNU General Public License along with Beebium.
# If not, see <https://www.gnu.org/licenses/>.

"""Python wrapper for ExtensionUiService.

The Extension UI framework lets any extension declare a typed control
tree (the *View*) which the server streams to clients. Clients render
natively and post user actions back through Dispatch. This module wraps
the gRPC SubscribeView / Dispatch RPCs and converts the proto types
into ergonomic dataclasses.

See ``docs/discussion/extension-ui-architecture.md`` for the design.

Usage::

    # One-shot: get the current view of the piconet panel.
    view = next(bbc.extension_ui.subscribe_view("piconet"))
    for control in view.root.children:
        print(control.id, control.kind)

    # Toggle the piconet "Enabled" control.
    bbc.extension_ui.dispatch("piconet", "mode_toggle", view.revision,
                               payload=False)

    # Background: receive view updates as they arrive.
    handle = bbc.extension_ui.start_background_subscription(
        "piconet", lambda v: print("revision", v.revision))
    ...
    handle.stop()
"""

from __future__ import annotations

import enum
import threading
from collections.abc import Callable, Iterator
from dataclasses import dataclass

from beebium.client._proto import extension_ui_pb2, extension_ui_pb2_grpc


class IndicatorState(enum.IntEnum):
    """Semantic state for an Indicator control.

    Renderers map this to whatever colour / icon convention their
    platform prefers; the server reports semantics, not presentation.
    """

    UNKNOWN = 0
    OK = 1
    WARN = 2
    ERROR = 3


class ControlKind(enum.Enum):
    """Tag for the seven control primitives plus the structural Group."""

    LABEL = "label"
    INDICATOR = "indicator"
    TOGGLE = "toggle"
    BUTTON = "button"
    CHOICE = "choice"
    TEXT_INPUT = "text_input"
    GROUP = "group"
    UNSET = "unset"


@dataclass(frozen=True)
class Label:
    text: str = ""
    # Optional muted caption shown alongside the primary text. Used
    # for provenance / status / size metadata that's logically
    # attached to the primary line. Empty means no caption.
    secondary_text: str = ""


@dataclass(frozen=True)
class Indicator:
    state: IndicatorState = IndicatorState.UNKNOWN
    text: str = ""


@dataclass(frozen=True)
class Toggle:
    label: str = ""
    value: bool = False


@dataclass(frozen=True)
class Button:
    label: str = ""
    enabled: bool = True


@dataclass(frozen=True)
class Choice:
    label: str = ""
    options: tuple[str, ...] = ()
    selected_index: int = 0


@dataclass(frozen=True)
class TextInput:
    label: str = ""
    value: str = ""
    placeholder: str = ""


@dataclass(frozen=True)
class Group:
    """A nested group of controls. ``label`` is None for unlabelled groups."""

    label: str | None = None
    children: tuple[Control, ...] = ()


@dataclass(frozen=True)
class Control:
    """A control in the View tree.

    Exactly one of ``label``, ``indicator``, ``toggle``, ``button``,
    ``choice``, ``text_input``, ``group`` is populated, indicated by
    the ``kind`` discriminator.
    """

    id: str
    kind: ControlKind
    label: Label | None = None
    indicator: Indicator | None = None
    toggle: Toggle | None = None
    button: Button | None = None
    choice: Choice | None = None
    text_input: TextInput | None = None
    group: Group | None = None


@dataclass(frozen=True)
class View:
    """A complete view of an extension's UI at one revision."""

    extension_id: str
    revision: int
    root: Control


def _convert_control(proto: extension_ui_pb2.Control) -> Control:
    """Convert a single proto Control into the dataclass form."""
    case = proto.WhichOneof("control")
    if case == "label":
        return Control(
            id=proto.id,
            kind=ControlKind.LABEL,
            label=Label(
                text=proto.label.text,
                secondary_text=proto.label.secondary_text,
            ),
        )
    if case == "indicator":
        return Control(
            id=proto.id,
            kind=ControlKind.INDICATOR,
            indicator=Indicator(
                state=IndicatorState(proto.indicator.state),
                text=proto.indicator.text,
            ),
        )
    if case == "toggle":
        return Control(
            id=proto.id,
            kind=ControlKind.TOGGLE,
            toggle=Toggle(
                label=proto.toggle.label,
                value=proto.toggle.value,
            ),
        )
    if case == "button":
        return Control(
            id=proto.id,
            kind=ControlKind.BUTTON,
            button=Button(
                label=proto.button.label,
                enabled=proto.button.enabled,
            ),
        )
    if case == "choice":
        return Control(
            id=proto.id,
            kind=ControlKind.CHOICE,
            choice=Choice(
                label=proto.choice.label,
                options=tuple(proto.choice.options),
                selected_index=proto.choice.selected_index,
            ),
        )
    if case == "text_input":
        return Control(
            id=proto.id,
            kind=ControlKind.TEXT_INPUT,
            text_input=TextInput(
                label=proto.text_input.label,
                value=proto.text_input.value,
                placeholder=proto.text_input.placeholder,
            ),
        )
    if case == "group":
        children = tuple(_convert_control(c) for c in proto.group.controls)
        # 'label' is an optional field in proto3; HasField distinguishes
        # "not set" from "set to empty string".
        label = proto.group.label if proto.group.HasField("label") else None
        return Control(
            id=proto.id,
            kind=ControlKind.GROUP,
            group=Group(label=label, children=children),
        )
    return Control(id=proto.id, kind=ControlKind.UNSET)


def _convert_view(proto: extension_ui_pb2.View) -> View:
    return View(
        extension_id=proto.extension_id,
        revision=proto.view_revision,
        root=_convert_control(proto.root),
    )


@dataclass(frozen=True)
class DispatchResult:
    """Return value from Extension UI dispatch."""

    accepted: bool
    error: str = ""


class SubscriptionHandle:
    """Handle for a background view subscription."""

    def __init__(self, thread: threading.Thread, stop_event: threading.Event):
        self._thread = thread
        self._stop_event = stop_event

    def stop(self, timeout: float = 1.0) -> None:
        """Signal the background subscription to stop and join the thread.

        This is a prompt cancel: the loop checks the stop flag before each
        callback, so an item already pulled from the stream when stop() is
        called is not delivered. To let a finite stream finish and deliver every
        item, use wait() instead.
        """
        self._stop_event.set()
        self._thread.join(timeout)

    def wait(self, timeout: float | None = None) -> bool:
        """Block until the subscription finishes on its own -- a finite stream
        ends, or the server closes it -- and report whether it did. Unlike
        stop(), this does not signal cancellation; it only waits."""
        self._thread.join(timeout)
        return not self._thread.is_alive()

    @property
    def is_running(self) -> bool:
        return self._thread.is_alive()


class ExtensionUi:
    """Wraps ExtensionUiService.

    Use ``subscribe_view`` to iterate over View updates for a named
    extension, and ``dispatch`` to post a user action back. The server
    delivers the response of an action as the next yielded View on the
    subscription, not in the dispatch return -- ``DispatchResult.accepted``
    only reports whether the framework's validation gauntlet (extension
    exists, control id known, payload type matches, revision current)
    passed.
    """

    def __init__(self, stub: extension_ui_pb2_grpc.ExtensionUiServiceStub):
        self._stub = stub

    def subscribe_view(self, extension_id: str) -> Iterator[View]:
        """Open a server-stream of View updates for one extension.

        The first View arrives immediately; subsequent Views arrive as
        the extension's state changes.

        Raises:
            grpc.RpcError: with code NOT_FOUND if the extension does
                not exist or has no UI.
        """
        request = extension_ui_pb2.SubscribeViewRequest(extension_id=extension_id)
        for proto_view in self._stub.SubscribeView(request):
            yield _convert_view(proto_view)

    def start_background_subscription(
        self,
        extension_id: str,
        callback: Callable[[View], None],
    ) -> SubscriptionHandle:
        """Subscribe in a daemon thread; invoke callback on each View.

        Returns a handle that can ``.stop()`` the background thread.
        Exceptions inside the stream (e.g. server disconnect, NOT_FOUND)
        terminate the thread silently -- check ``handle.is_running`` if
        the caller needs to detect that.
        """
        stop_event = threading.Event()

        def loop() -> None:
            try:
                for view in self.subscribe_view(extension_id):
                    if stop_event.is_set():
                        break
                    callback(view)
            except Exception:
                pass

        thread = threading.Thread(target=loop, daemon=True)
        thread.start()
        return SubscriptionHandle(thread, stop_event)

    def dispatch(
        self,
        extension_id: str,
        control_id: str,
        view_revision: int,
        payload: bool | str | int | None = None,
    ) -> DispatchResult:
        """Post a user action back to the server.

        The ``payload`` is type-dispatched to match the addressed
        control's expected variant:

            * ``bool``  -> Toggle
            * ``str``   -> TextInput
            * ``int``   -> Choice (selected index, must be non-negative)
            * ``None``  -> Button (no payload)

        Returns ``DispatchResult(accepted=False, error=...)`` if the
        framework's validation gauntlet rejects the request (stale
        revision, unknown control, payload-type mismatch, unknown
        extension). The actual mutation of the extension's state is
        observable on the SubscribeView stream as the next pushed View,
        not in the return value.
        """
        request = extension_ui_pb2.DispatchRequest(
            extension_id=extension_id,
            control_id=control_id,
            view_revision=view_revision,
        )
        if payload is None:
            pass  # leave the oneof unset (Button)
        elif isinstance(payload, bool):
            request.bool_value = payload
        elif isinstance(payload, str):
            request.string_value = payload
        elif isinstance(payload, int):
            if payload < 0:
                raise ValueError(f"Choice index payload must be non-negative; got {payload}")
            request.index_value = payload
        else:
            raise TypeError(f"dispatch payload must be bool, str, int, or None; got {type(payload).__name__}")

        response = self._stub.Dispatch(request)
        return DispatchResult(accepted=response.accepted, error=response.error)
